File: jun1/src/combinators/eps.py
from ..combinator import Combinator
from ..parse_iteration_result import FrameStack, ParserIterationResult
from ..state import CombinatorState
from ..u8set import U8Set


class EpsState(CombinatorState):
    def __init__(self, frame_stack: FrameStack) -> None:
        self.frame_stack = frame_stack


class Eps(Combinator):
    def initial_state(self, signal_id: int, frame_stack: FrameStack) -> EpsState:
        return EpsState(frame_stack)

    def next_state(self, state: EpsState, c: Optional[str], signal_id: int) -> tuple[EpsState, int, ParserIterationResult]:
        return state, signal_id, ParserIterationResult(U8Set.none(), True, state.frame_stack)

File: jun1/src/combinators/eps.rs
use crate::combinator::Combinator;
use crate::parse_iteration_result::{FrameStack, ParserIterationResult};
use crate::state::CombinatorState;
use crate::U8Set;

pub struct Eps;

impl Combinator for Eps {
    type State = EpsState;

    fn initial_state(&self, _signal_id: &mut usize, frame_stack: FrameStack) -> Self::State {
        EpsState { frame_stack }
    }

    fn next_state(&self, state: &mut Self::State, _c: Option<char>, _signal_id: &mut usize) -> ParserIterationResult {
        let state = state.as_any_mut().downcast_mut::<EpsState>().expect("Invalid state type");
        let mut result = ParserIterationResult::new(U8Set::none(), true, state.frame_stack.clone());
        result
    }
}

pub struct EpsState {
    pub frame_stack: FrameStack,
}

impl CombinatorState for EpsState {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}
